src/probnum/diffeq/odefiltsmooth/diffusions.py

```diff
@@ -0,0 +1,120 @@
+from probnum.type import FloatArgType
+
+DiffusionType = Union[FloatArgType]
+"""Acceptable diffusion types are -- in principle -- floats and arrays of shape (d,).
+In other words, everything that behaves well with `cov1 + diffusion*cov2` and
+`chol1 + sqrt(diffusion) * chol2`.
+For now, let's only use floats, because it is not clear to me how to best implement the square-root behaviour.
+"""
+
+
+class Diffusion(abc.ABC):
+    r"""Interface for diffusion models :math:`\sigma: \mathbb{R} \rightarrow \mathbb{R}^d` and their calibration."""
+
+    def __repr__(self):
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def __call__(self, t) -> DiffusionType:
+        """Evaluate the diffusion."""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def __getitem__(self, idx):
+        """Get the i-th diffusion from the list."""
+        pass
+
+    def calibrate_locally(self, meas_rv):
+        """Compute a local diffusion estimate.
+
+        Used for uncertainty calibration in the ODE solver.
+        """
+        std_like = meas_rv.cov_cholesky
+        whitened_res = np.linalg.solve(std_like, meas_rv.mean)
+        ssq = whitened_res @ whitened_res / meas_rv.size
+        return ssq
+
+    @abc.abstractmethod
+    def update_current_information(self, diffusion, t):
+        """Update the current information about the global diffusion.
+
+        This could mean appending the diffusion to a list or updating a
+        global estimate.
+        """
+        pass
+
+    @abc.abstractmethod
+    def postprocess_states(self, states, locations):
+        """Postprocess a set of ODE solver states after seeing all the data."""
+        pass
+
+
+class ConstantDiffusion(Diffusion):
+    """Piecewise constant diffusion and their calibration."""
+
+    def __init__(self):
+        self.diffusion = None
+        self._seen_diffusions = 0
+
+    def __repr__(self):
+        return f"ConstantDiffusion({self.diffusion})"
+
+    def __call__(self, t) -> DiffusionType:
+        return self.diffusion
+
+    def __getitem__(self, idx):
+        return self.diffusion
+
+    def update_current_information(self, diffusion, t):
+        """Update the current global MLE with a new diffusion."""
+        self._seen_diffusions += 1
+
+        if self.diffusion is None:
+            self.diffusion = diffusion
+        else:
+            a = 1 / self._seen_diffusions
+            b = 1 - a
+            self.diffusion = a * diffusion + b * self.diffusion
+
+    def postprocess_states(self, states, locations):
+        return [
+            random_variables.Normal(rv.mean, self.diffusion * rv.cov) for rv in states
+        ]
+
+
+class PiecewiseConstantDiffusion(Diffusion):
+    """Piecewise constant diffusion."""
+
+    def __init__(self):
+        self.diffusions = []
+        self.times = []
+
+    def __repr__(self):
+        return f"PiecewiseConstantDiffusion({self.diffusions})"
+
+    def __call__(self, t) -> DiffusionType:
```

Comment:
Why should we not implement this, too? That way, the diffusion would behave like a real step function, and it should be possible since the diffusion object keeps track of `times` already.

Reply:
true. I have not done it yet, because it is not needed anywhere yet and because I wanted to keep this PR as small as possible. Do you want me to include this here? I can do it.

Reply:
If it's fine for you, I think for the sake of completeness it should ideally be implemented as part of this PR. Thanks!